### ISSUE TITLE
Failed to fetch JSON data

### ISSUE THREAD

**Issue by bodyguardroman** [matthuisman/slyguy.addons#140]
Hi,iam from Europe(Slovakia)Hbo max launched into my country.Its working on official app,but in kodi its write:make sure its your ip adress is allowed.will be work later,or its addon only for US ip adress?

**Comment by matthuisman** [comment]
Should work everywhere.
Please provide full debug log

**Comment by bodyguardroman** [comment]
Kodi log:
https://paste.kodi.tv/azifamojog.kodi
Error hbo max included(normal i can signing,but i can't play video. i dont unterstand what happenned,other addons working...

**Comment by matejmosko** [comment]
It's same for me (Slovakia) - login is without problems. Playing videos ends with "Failed to fetch JSON data". @bodyguardroman have you done anything to solve the issue or it just fixed by itself?

**Comment by bodyguardroman** [comment]
> It's same for me (Slovakia) - login is without problems. Playing videos ends with "Failed to fetch JSON data". @bodyguardroman have you done anything to solve the issue or it just fixed by itself?

I dont know where is mistake,but i didnt do anything.starting working,but probably,its new.Maybe it will be long time,when starting normal working full database...

**Comment by matthuisman** [comment]
@matejmosko 
do you still have issue?

**Comment by matejmosko** [comment]
It's working now. Meanwhile I found out there was an email asking me to verify my email address. I have done it. I am not quite sure if the verification helped or the time solved it. Playing videos through official android app worked even without verification.

**Comment by matthuisman** [comment]
oh right, so you were trying to login to the kodi add-on before verifying your email address?
Interesting that the official app worked without that

**Comment by bszente** [comment]
Hi @matthuisman !

I can confirm the `Failed to fetch JSON data` error, I'm from Romania.

Steps I did:

1. Yesterday I installed the plugin in my Linux based media box.
2. I did the `/tvsignin` login.
3. Yesterday everything went well, I could watch moves.
4. Today I receive this error when I want to navigate between the movies. I did not do any changes to Kodi/settings whatsoever.
5. The plugin starts, but when I want to enter into any category, I get the following message:
```
Failed to fetch JSON data
Make sure your IP address is allowed (not geo-blocked)
Try accessing the content via other means (offical app / website) to test if it's a Kodi add-on issue
```
6. I tried HBO max from my laptop and it works. So there is no geo-blocking.

It's like the obtained authentication token would be expired, or something like that.

Please find attached the relevant part of the Debug log.
[debug_logs.txt](https://github.com/matthuisman/slyguy.addons/files/8225000/debug_logs.txt)

**Comment by bszente** [comment]
I did a `Logout` and afterwards a `Login` again, but this time with email and password. I do not get the error anymore.

Is there any difference between the TV token and email/pass login type?

**Comment by Szurgent** [comment]
I have some issue.. i think..

1. I log in with my email/pass and watch some movies...
2. next day i received the "Failed to fetch JSON data..." error.
3. i cant login again
4. i change my pass, and log with device (i use https://activate.hbomax.com/) ... watch movies.
5. Now i received (after 9 hours) the "Failed to fetch JSON data..."???

**Comment by bszente** [comment]
I'm back to confirm, that after a few hours I get the `Failed to fetch JSON data` again. So the issue is not related to TV Token vs. email/pass login type.

It seems that the authentication token gets invalidated or something like that.

If you need some extra logs, info, please let us know, so we can help. Thank you very much for your valuable work @matthuisman !

**Comment by matthuisman** [comment]
yup, seems when the token expires - its not refreshing correctly.
ill try get fixed today :)

**Comment by matthuisman** [comment]
ok, v0.9.2 should sort it out hopefully :)

please everyone try and let me know

**Comment by Szurgent** [comment]
It's Working . The error disappered withouth relogin for me. thx the quick fix. My 3-year-old son is also grateful.

**Comment by bszente** [comment]
I confirm 0.9.2 works! Thank you very much for the quick fix!

**Comment by rbnmndz** [comment]
Thank you!

**Comment by samukas** [comment]
Working for me as well. Thank you !

**Comment by samukas** [comment]
So... I've had this error show up again, I'm afraid.
Unlike last time, I am able to navigate all menus, the JSON error just shows up after triggering a "play" action.

Please find attached the relevant part of the Debug log (profile ID obfuscated)
[debug_json.txt](https://github.com/matthuisman/slyguy.addons/files/8239332/debug_json.txt)

**Comment by samukas** [comment]
Interesting. Now, 5-6 hours later it's back to working again. I didn't change anything in Kodi or the add-on, I didn't sign-out / sign-in, etc.

Pure speculation here, but could it be that the token is not being refreshed as "early" as needed and that there are gaps during the day where the add-on is not working? Hum... I'll keep checking and report here if it happens again.

**Comment by matthuisman** [comment]
Think maybe it has to do with the caching of the config. When the token expires, I don't think I'm clearing the config cache. Could maybe be that. Either that or the expires time they advise is not accurate

**Comment by bszente** [comment]
@samukas can you confirm please, if your system clock is set properly (e.g. do you have NTP sync active)?

**Comment by samukas** [comment]
I couldn't find a setting for "NTP Sync", but I'm guessing Kodi might be grabbing date and time from Android?
I'm using an Nvidia Shield - which is, indeed, set to automatic and the right timezone.

I have less time at home during the week to keep testing, but I have to confess, I haven't been able to reproduce this since yesterday's error. I'm kicking myself for not remembering to have a look yesterday at the addon data xml when the error was ocurring, that could have been revealing.

The only two scenarios I can think of at the moment are:
- Somehow, an invalid "access_token" was generated and/or stored. And because of that, I was stuck with it until it was time again for a refresh, i.e., when what was stored under "userdata" -> "expires" had elapsed . Which, in the case of my account, appears to be every 4 hours.
- A cache issue, as @matthuisman mentioned. Although so far, I've done several attempts to play and the addon data xml seems to overwrite when needed.
 
Weirdly enough though, "refresh_token" and "access_token" seem to be 100% equal each time, which I thought was odd ?

**Comment by secretcreature** [comment]
Hello, I also have this issue (Failed to fetch JSON data), my addon version is 0.9.2. 
During the day addon was working correctly, but in the evening mentioned error ocurred. I tried logout and then login - error disappeared. Based on this it looks like issue related to token validity.

**Comment by samukas** [comment]
Just thought I'd add that, personally, after witnessing this issue on March 13, I haven't had it reappear since then

**Comment by secretcreature** [comment]
It happens to me almost every day, I have to logout/login or click "reset addon" button, then login and it works...

**Comment by matthuisman** [comment]
just pushed 0.9.3
lets see how that goes for the json error.
it now makes sure to clear the config cache before refreshing token

**Comment by secretcreature** [comment]
@matthuisman thanks for the new version, it looks like it works, no more json error since upgrade...

**Comment by matthuisman** [comment]
interesting, ill investigate

also, i didnt know you had updated the comment - so was thinking this was fine
please post new comments if there is a change in status so i get notifed :)

**Comment by secretcreature** [comment]
OK, I will do that, but it is not possible for me to replicate that error, so we have to wait for the error to pop up :)
With version 0.93 json error occurred only once, so it is a huge improvement regarding this issue.
Many thanks for your effort.

**Comment by band1959** [comment]
Hi @matthuisman !
I have the foxtel go installed and has been working well but today all of a sudden today it says failed to fetch JSON data and I am in Australia. Any idea why this is happening?
Thanks Ben

**Comment by matthuisman** [comment]
Why would you put that on this HBO issue?

Please create a new issue with full debug log

**Comment by band1959** [comment]
sorry new here and not sure but will research how to get debug log

**Comment by secretcreature** [comment]
no, no more errors like that, thanks. I do not watch hbo too much, but my kids did not complain, so I assume it is fixed. 
Thanks for your support.

**Comment by jarnehed** [comment]
I am having this error again as of Dec 9th 2023. Located in Sweden.
